File: LeetCode/3499_Maximize_Active_Section_with_Trade_I.swift
class Solution {
    func maxActiveSectionsAfterTrade(_ s: String) -> Int {
        let runs = s.chunked(by: { $0 == $1 }).map(\.count)

        let startsWithOne = s.first == "1"
        let isOneBlock = { (i: Int) in startsWithOne ? i % 2 == 0 : i % 2 == 1 }

        let result = runs.enumerated()
            .filter { isOneBlock($0.offset) }
            .reduce(0) { $0 + $1.element }

        let zeroBlocks = runs.enumerated()
            .filter { !isOneBlock($0.offset) }
            .map(\.element)

        let maxGain = zeroBlocks
            .adjacentPairs()
            .map { $0 + $1}
            .max() ?? 0

        return result + maxGain
    }
}

File: LeetCode/3499_Maximize_Active_Section_with_Trade_I.cpp
class Solution {
public:
    int maxActiveSectionsAfterTrade(string s) {
        // Run-length encoding: even index = 1-blocks, odd index = 0-blocks
        vector<int> runs = {0};
        for (char c : s) {
            if ((c - '0') != (int)runs.size() % 2) {
                runs.push_back(0);
            }
            runs.back()++;
        }
        if (s.back() == '0') runs.push_back(0);

        int result = 0;
        int windowSum = 0;
        int maxGain = 0;

        for (int i = 0; i < (int)runs.size(); i++) {
            if (i % 2 == 0) {
                result += runs[i];
            } else {
                windowSum += runs[i];
                if (i >= 4) windowSum -= runs[i - 4];
                if (i >= 3) maxGain = max(windowSum, maxGain);
            }
        }

        return result + maxGain;
    }
};
